src/ccs.rs

```diff
@@ -0,0 +1,636 @@
+pub struct CCSShape<G: Group> {
+  pub(crate) num_cons: usize,
+  pub(crate) num_vars: usize,
+  pub(crate) num_io: usize,
+
+  pub(crate) M: Vec<SparseMatrix<G>>,
+  pub(crate) t: usize,
+  pub(crate) q: usize,
+  pub(crate) d: usize,
+  pub(crate) S: Vec<Vec<usize>>,
+
+  // Was: usize
+  pub(crate) c: Vec<G::Scalar>,
+
+  // m is the number of columns in M_i
+  pub(crate) m: usize,
+  // n is the number of rows in M_i
+  pub(crate) n: usize,
+  // s = log m
+  pub(crate) s: usize,
+  // s_prime = log n
+  pub(crate) s_prime: usize,
+}
+
+/// A type that holds a witness for a given CCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct CCSWitness<G: Group> {
+  // Vector W in F^{n - l - 1}
+  W: Vec<G::Scalar>,
+}
+
+// TODO: Make sure this is in the right form for committed CCS using MLE, possibly a separate type
+/// A type that holds an CCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCSInstance<G: Group> {
+  // XXX: Move commitment out of CCSInstance for more clean conceptual separation?
+  // (Pedersen) Commitment to a witness
+  pub(crate) comm_W: Commitment<G>,
+
+  // Public input x in F^l
+  pub(crate) X: Vec<G::Scalar>,
+}
+
+/// A type that holds the shape of a Committed CCS (CCCS) instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCCSShape<G: Group> {
+  // Sequence of sparse MLE polynomials in s+s' variables M_MLE1, ..., M_MLEt
+  // TODO This should be MLE, but possible we don't have to keep in struct?
+  // Exists in paper but not multifolding-poc
+  pub(crate) M_MLE: Vec<SparseMatrix<G>>,
+
+  // XXX Embed CCS directly here or do a flat structure?
+  pub(crate) ccs: CCSShape<G>,
+  // q multisets S (same as CCS)
+  // q constants c (same as CCS)
+}
+
+/// CCCS Instance is (C, x)
+/// CCCS Witness is w _mle
+
+// NOTE: We deal with `r` parameter later in `nimfs.rs` when running `execute_sequence` with `ro_consts`
+/// A type that holds a CCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct CCCSInstance<G: Group> {
+  // Commitment to a multilinear polynomial in s' - 1 variables
+  pub(crate) C: Commitment<G>,
+
+  // $x in F^l$
+  pub(crate) X: Vec<G::Scalar>,
+}
+
+/// A type that holds a witness for a given CCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct CCCSWitness<G: Group> {
+  // Multilinear polynomial w_mle in s' - 1 variables
+  pub(crate) w_mle: Vec<G::Scalar>,
+}
+
+// NOTE: We deal with `r` parameter later in `nimfs.rs` when running `execute_sequence` with `ro_consts`
+/// A type that holds a LCCCS instance
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(bound = "")]
+pub struct LCCCSInstance<G: Group> {
+  pub(crate) C: Commitment<G>,
+  pub(crate) X: Vec<G::Scalar>,
+  pub(crate) u: G::Scalar,
+  pub(crate) v: Vec<G::Scalar>,
+}
```

Comment:
As for `CCCSShape`, this will also need access to the underlying `CCSShape` structure.